system/codesigner/gpgcodesigner.cpp

```diff
@@ -0,0 +1,278 @@
+#include "jlib.hpp"
+#include "jexcept.hpp"
+#include "jlog.hpp"
+#include "gpgcodesigner.hpp"
+
+class GpgCodeSigner : implements ICodeSigner
+{
+    virtual bool sign(const char * text, const char * userId, const char * passphrase, StringBuffer & errmsg, StringBuffer & signedText) override;
+    virtual bool verifySignature(const char * text, StringBuffer & signer, StringBuffer & errmsg) override;
+    virtual bool hasSignature(const char * text) const override;
+    virtual StringBuffer &stripSignature(const char * text, StringBuffer & unsignedText) const override;
+    virtual StringArray &getUserIds(StringArray & userIds, StringBuffer & errmsg) override;
+private:
+    void initGpg(void);
+    bool getKeyGrip(const char * user, StringBuffer & keygrip);
+    void clearPassphrase(const char * key);
+    StringBuffer & sanitizeUserId(const char * userId, StringBuffer & sanitizedUserId);
+    StringBuffer & sanitizePassphrase(const char * passphrase, StringBuffer & sanitizedPassphrase);
+    bool isGpgV1 = false;
+    bool isGpgInstalled = false;
+    static constexpr const char* signatureMsgHeader = "-----BEGIN PGP SIGNED MESSAGE-----";
+    static constexpr const char* signatureBegin = "-----BEGIN PGP SIGNATURE-----";
+    CriticalSection crit;
+};
+
+GpgCodeSigner gpgCodeSigner;
+
+extern jlib_decl ICodeSigner &queryGpgCodeSigner()
+{
+    return gpgCodeSigner;
+}
+
+void GpgCodeSigner::initGpg(void)
+{
+    if (isGpgInstalled) return;
+
+    CriticalBlock block(crit);
+    StringBuffer cmd, output, errmsg;
```

Comment:
isGpgInstalled should be checked again within critical block. Also strictly it should be an atomic variable.

Reply:
Done.